pod: add exit policies

vrothberg opened the pull request:
Add the notion of an "exit policy" to a pod.  This policy controls the
behaviour when the last container of pod exits.  Initially, there are
two policies:

 - "continue" : the pod continues running. This is the default policy
                when creating a pod.

 - "stop" : stop the pod when the last container exits. This is the
            default behaviour for `play kube`.

In order to implement the deferred stop of a pod, add a worker queue to
the libpod runtime.  The queue will pick up work items and in this case
helps resolve dead locks that would otherwise occur if we attempted to
stop a pod during container cleanup.

Note that the default restart policy of `play kube` is "Always".  Hence,
in order to really solve #13464, the YAML files must set a custom
restart policy; the tests use "OnFailure".

Fixes: #13464
Signed-off-by: Valentin Rothberg <vrothberg@redhat.com>

vrothberg commented on libpod/pod_api.go:
```diff
@@ -195,6 +200,49 @@ func (p *Pod) StopWithTimeout(ctx context.Context, cleanup bool, timeout int) (m
+
+	for _, ctr := range allCtrs {
+		if ctr.ID() == infra.ID() {
+			// Skip the receiver (dead lock) and the infra container.
```
```suggestion
```
From an earlier version

mheon replied:
What does this mean? I can't really parse this comment

cdoern replied:
Does this skip the receiver or just infra? I ran into the same issue here where podman would deadlock when trying to stop both the container calling and infra...

vrothberg replied:
We need to skip checking the infra container here as it is still running; we'd otherwise return early below and not stop it.

mheon commented:
I don't think this works with `--rm` containers, which will directly invoke `podman rm` last I checked.

vrothberg commented on libpod/container_internal.go:
```diff
@@ -1927,6 +1927,25 @@ func (c *Container) cleanup(ctx context.Context) error {
+				lastError = fmt.Errorf("container %s is in pod %s, but pod cannot be retrieved: %w", c.ID(), c.config.Pod, err)
+			}
+		} else {
+			c.runtime.queueWork(func() {
```
It's happening in another go routine which would then be blocked until this one here returns from `cleanup()`.

vrothberg replied:
The idea is to have a caller outside this stack to prevent abba locks and the need for recursive locks etc.

mheon replied:
Sure - but we're waiting for the error return from the function here? I assumed we blocked waiting for it to execute, am I incorrect?

How do we guarantee the worker finishes before Podman exits?

Luap99 replied:
Yes I also don't understand this. If the problem are deadlocks we should fix the underlying code path.

vrothberg replied:
We're not waiting. We queue work which will happen at some later point.

vrothberg replied:
> If the problem are deadlocks we should fix the underlying code path.

We just can't. Assume one process holds the lock of pod p1, and wants to list all containers and tries to lock container c1. Now container c1 exits, locks c1 before the other process can and tries to do what we're doing ... at some point it needs to lock pod p1. ABBA dead lock.

There is much more potential for other dead locks.

vrothberg replied:
>  I assumed we blocked waiting for it to execute, am I incorrect?

Yes, this assumption is incorrect. `queueWork()` returns immediately and cannot block.

mheon commented:
Actually, I take that back, `RemoveContainer()` should invoke `cleanup()` if the container was running.

vrothberg commented on libpod/container_internal.go:
```diff
@@ -1927,6 +1927,25 @@ func (c *Container) cleanup(ctx context.Context) error {
 		}
 	}
 
+	if c.config.Pod != "" {
```
I was considering that but was afraid to rely on all callers to change the state before calling `cleanup()`.   Note that the pod won't get stopped if this or any container is another state than exited, removing or stopped.  So the security net to not make something stupid is in `stopInfraIfNeeded()`.

mheon commented:
We can potentially avoid the workqueue bits by doing this in a `defer()` in `Cleanup()` and `RemoveContainer()` - as long as it's the first defer in the function, even before the lock/defer unlock() - it will run after all locks are released, but still block the function from returning and Podman from exiting until it finishes.

vrothberg commented on libpod/pod_api.go:
```diff
@@ -195,6 +200,49 @@ func (p *Pod) StopWithTimeout(ctx context.Context, cleanup bool, timeout int) (m
+		return nil
+	}
+
+	infra, err := p.infraContainer()
```
Nice idea, thanks!

mheon commented:
Though, that requires that we add it to every function that can stop a single container, which could be complicated

vrothberg commented:
> Note my use of the capitalized `Cleanup` - if we hit the public API entrypoints, we can avoid all locks because no locks have been taken before `Cleanup()` starts

How would you call `stopInfraIfNeeded` after the API finished?  I guess with a `defer` and that's where the container is locked.

vrothberg commented:
> We could make the auto-stop behavior opt-in and let play kube default to it.

Given e2e tests are also barking, I think the behavior should really be opt-in for ordinary pods but be turned on for `play kube` ones.

Luap99 commented:
> Given e2e tests are also barking, I think the behavior should really be opt-in for ordinary pods but be turned on for `play kube` ones.

I can only speak for the `podman run check dnsname plugin with CNI` test, in this case it is just a implementation detail and should be fixed. It tries to resolve the pod name because it assumes it stays running, adding another container to keep it running will fix this. I don't see this as a real world use case.

> We could make the auto-stop behavior opt-in and let play kube default to it.

I think making this configurable is a good idea in case users have problems with the new bahavior but IMO the default should be to stop the infra. I see no reasons why they have to keep running since they do nothing.

rhatdan commented:
I thought the use case we added or were adding was pods go away when the last container exits.  They can start without a container, but if a container is added and removed the infra container goes away.

vrothberg commented:
> I thought the use case we added or were adding was pods go away when the last container exits. They can start without a container, but if a container is added and removed the infra container goes away.

The infra container will now be stopped but pod will continue to exist.  We can add a `--rm` flag to pod-create similar to the one for containers to remove on exit.

The question now is whether we want to auto-stop the infra container for ordinary pods.  There are certain tests breaking that created a file in the pod's mount NS:
* First container does a `touch /foo`
* Second container does a `ls /foo`

The second one is now breaking as the mount NS is destroyed when we stop the infra container.

rhatdan commented:
Can we change the test to do:

```
podman create mypod -v test:/test --name mypod
podman pod run --name mypod $IMAGE touch /test/foobar
podman pod run --name mypod $IMAGE ls /test/foobar
```
Second container should restart the pod, correct?

vrothberg commented on libpod/container_internal.go:
```diff
@@ -1939,9 +1939,41 @@ func (c *Container) cleanup(ctx context.Context) error {
 	}
 
+	if err := c.stopPodIfNeeded(context.Background()); err != nil {
+		return err
```
Should do the `lastError` dance as above.

rhatdan commented on cmd/podman/pods/create.go:
```diff
@@ -72,6 +72,9 @@ func init() {
 	flags.StringVarP(&createOptions.Name, nameFlagName, "n", "", "Assign a name to the pod")
 	_ = createCommand.RegisterFlagCompletionFunc(nameFlagName, completion.AutocompleteNone)
 
+	flags.StringVarP(&createOptions.ExitPolicy, "exit-policy", "", "continue", "Behaviour when the last container exits")
```
This is fine for now, but this should be set in containers.conf.

Luap99 replied:
nit make "exit-policy" a var instead of duplicating it twice.

rhatdan commented:
LGTM

mheon commented:
I'm still unsure how we guarantee all pending jobs have finished before the runtime shuts down and Podman exits

TomSweeneyRedHat commented:
All kinds of red unhappy test and you need a rebase @vrothberg

vrothberg commented:
> I'm still unsure how we guarantee all pending jobs have finished before the runtime shuts down and Podman exits

I'll take another look

vrothberg commented:
@mheon, can you take a look at the code comments and I added? Please let me know if you find them sufficient.

Luap99 commented on docs/source/markdown/podman-pod-create.1.md:
```diff
@@ -75,6 +75,13 @@ Set custom DNS options in the /etc/resolv.conf file that will be shared between
 
 Set custom DNS search domains in the /etc/resolv.conf file that will be shared between all containers in the pod.
 
+#### **--exit-policy**=*{continue,stop}*
```
I know everyone ignores that but we have a https://github.com/containers/podman/blob/main/docs/MANPAGE_SYNTAX.md

and the correct syntax is
```suggestion
#### **--exit-policy**=**continue** | *stop*
```

Luap99 commented on libpod/define/pod_exit_policy.go:
```diff
@@ -0,0 +1,34 @@
+import "fmt"
+
+// PodExitPolicies includes the supported pod exit policies.
+var PodExitPolicies = []string{"continue", "stop"}
```
do not duplicate the strings please, use `string(PodExitPolicyContinue)` and `string(PodExitPolicyStop)`

vrothberg commented:
Moving to WIP for now to do the plumbing for containers.conf.

vrothberg commented:
-> https://github.com/containers/common/pull/1017 to move the definitions and parsing directly to c/common.

flouthoc commented:
@vrothberg Just a small doubt why do we need extra `exit_policy` shouldn't pod just follow how `restartPolicy` is configured default is `Always` so it is expected to restart but its `Never` or `OnFailure` pod should just end with `Succeeded` status.

rhatdan commented:
Containers follow restart policy not the Pod.  Theoretically if I add a restart=always container to a Pod, the container should restart even if the Pod is set to `stop`,

flouthoc commented:
@rhatdan ah i see but shouldn't containers inherit the `restartPolicy` from pod spec as stated here for all the containers in a pod https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy

Afaik In kubernetes construct i don't think its possible to add a restart policy for a specific container , containers always inherit restartPolicy from pod , some reference here https://pkg.go.dev/k8s.io/api/core/v1#Container and https://pkg.go.dev/k8s.io/api/core/v1#PodSpec

Just want to make sure that a standard spec behaves in a similar manner between podman and k8s other than that I think this should be fine.

vrothberg commented:
> @vrothberg Just a small doubt why do we need extra `exit_policy` shouldn't pod just follow how `restartPolicy` is configured default is `Always` so it is expected to restart but if its `Never` or `OnFailure` pod should just end with `Succeeded` status.

It won't without this patch as the infra container would continue to run.  Hence, we need to stop the pod once all containers (other than the infra) are done. _However_, we cannot do that unconditionally as it would break backwards compat to how `podman pod create` works.  That's why we needed to make it configurable via an "exit policy".

vrothberg commented:
> Moving to WIP for now to do the plumbing for containers.conf.

Done.

vrothberg commented:
@mheon PTanotherL

Luap99 commented on cmd/podman/pods/create.go:
```diff
@@ -72,6 +72,10 @@ func init() {
 	_ = createCommand.RegisterFlagCompletionFunc(nameFlagName, completion.AutocompleteNone)
 
+	policyFlag := "exit-policy"
+	flags.StringVarP(&createOptions.ExitPolicy, policyFlag, "", string(containerConfig.Engine.PodExitPolicy), "Behaviour when the last container exits")
```
```suggestion
	flags.StringVar(&createOptions.ExitPolicy, policyFlag, string(containerConfig.Engine.PodExitPolicy), "Behaviour when the last container exits")
```
small nit in case you have to repush

Luap99 reviewed (commented):
LGTM

mheon commented:
LGTM on my end

vrothberg commented:
Nice, @rhatdan, want to merge?

rhatdan commented:
/lgtm
